src/managed/Microsoft.Extensions.DependencyModel/DependencyContextJsonReader.Netstandard2_0.cs

```diff
@@ -0,0 +1,628 @@
+                    case DependencyContextStrings.ServiceablePropertyName:
+                        serviceable = reader.ReadAsNullableBoolean(defaultValue: false);
+                        break;
+                    case DependencyContextStrings.PathPropertyName:
+                        path = reader.ReadAsString();
+                        break;
+                    case DependencyContextStrings.HashPathPropertyName:
+                        hashPath = reader.ReadAsString();
+                        break;
+                    case DependencyContextStrings.RuntimeStoreManifestPropertyName:
+                        runtimeStoreManifestName = reader.ReadAsString();
+                        break;
+                    default:
+                        reader.Skip();
+                        break;
+                }
+            }
+
+            reader.CheckEndObject();
+
+            return new LibraryStub()
+            {
+                Hash = hash,
+                Type = Pool(type),
+                Serviceable = serviceable,
+                Path = path,
+                HashPath = hashPath,
+                RuntimeStoreManifestName = runtimeStoreManifestName
+            };
+        }
+
+        private List<RuntimeFallbacks> ReadRuntimes(ref Utf8JsonReader reader)
+        {
+            var runtimeFallbacks = new List<RuntimeFallbacks>();
+
+            reader.ReadStartObject();
+
+            while (reader.Read() && IsTokenTypeProperty(reader.TokenType))
+            {
+                string runtime = reader.GetString();
+                var fallbacks = reader.ReadStringArray();
+
+                runtimeFallbacks.Add(new RuntimeFallbacks(runtime, fallbacks));
+            }
+
+            reader.CheckEndObject();
+
+            return runtimeFallbacks;
+        }
+
+        private static ArraySegment<byte> ReadToEnd(Stream stream)
+        {
+            int written = 0;
+            byte[] rented = null;
+
+            try
+            {
+                long expectedLength = 0;
+
+                if (stream.CanSeek)
+                {
+                    expectedLength = Math.Max(1L, stream.Length - stream.Position);
+                    rented = ArrayPool<byte>.Shared.Rent(checked((int)expectedLength));
+                }
+                else
+                {
+                    rented = ArrayPool<byte>.Shared.Rent(UnseekableStreamInitialRentSize);
+                }
+
+                int lastRead;
+
+                do
+                {
+                    if (expectedLength == 0 && rented.Length == written)
+                    {
+                        byte[] toReturn = rented;
+                        rented = ArrayPool<byte>.Shared.Rent(checked(toReturn.Length * 2));
+                        Buffer.BlockCopy(toReturn, 0, rented, 0, toReturn.Length);
+                        // Holds document content, clear it.
+                        ArrayPool<byte>.Shared.Return(toReturn, clearArray: true);
```

Comment:
What happens if the stream grows after its Length was initially checked in the CanSeek section above?  Won't we start calling stream.Read with invalid arguments?  For robustness, you might want to just treat the initial sizing as a best guess rather than guaranteed.

Reply:
I am borrowing this helper from corefx. Should it be fixed there as well?
https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs#L176

Reply:
> For robustness, you might want to just treat the initial sizing as a best guess rather than guaranteed.

I don't follow what you mean here. Can you elaborate what you recommend we do with the initial size (by which I am assuming you mean `expectedLength`)?

Reply:
I mean right now this code will blow up if the stream is seekable but its length changes over the course of the method.  I'm suggesting instead that you still presize the array based on the initially read Length, but then don't do anything differently in the rest of the method than for the !CanSeek case.

Reply:
> You are suggesting I remove the resize logic all together.

No, the opposite, that you always use the resize logic even if CanSeek == true... basically remove the "expectedLength == 0 &&" from the if check.  In the CanSeek case, you'd also want to ask for an array that was Length+1 rather than Length, so that you'd never hit this resize in the common case.

Reply:
> Can the stream length shrink as well?

Yes:
```C#
using System;
using System.IO;

class Program
{
    static void Main()
    {
        using (Stream theStream = new FileStream("tmp.dat", FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite))
        {
            Console.WriteLine(theStream.Length);
            ChangeLength("tmp.dat", 10);
            Console.WriteLine(theStream.Length);
            ChangeLength("tmp.dat", 5);
            Console.WriteLine(theStream.Length);
        }
    }

    static void ChangeLength(string path, int length)
    {
        using (Stream tmp = new FileStream(path, FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite))
            tmp.SetLength(length);
    }
}
```
outputs:
```
0
10
5
```

Reply:
The check against zero was to avoid the seekable stream resizing when it didn't change.  Doing `+ 1` and letting the resize run if it hits seems fine.

(The hint could also subtract off Position if it wanted more accuracy :smile:)